### PR TITLE
Dolphin, Defence, Cleanup

### PR DESCRIPTION
**Pull request by AdamMutzell**
I'm so sorry for this enormous commit and pull request. Been working very sporradically during the last few days, and could not get it structured. 

Anyways, heres a breakdown:
- Dolphin attack exists and has its own environment and training.
- Dolphin attack integrated into benchmarks and defence.
- Defence now split into two separate files. Both are heavily modified.
- Step functions now use perform_attack, like we talked about.
- Environment and training files cleaned and shortened. 

I think I've tested everything, but let me know if anything is broken.

### CODE REVIEW

File: audio/audio.py
Comment (Uplink036) [inline audio/audio.py:1]: I like the refactoring.

File: defence/classify_detect.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
+import librosa
+import numpy as np
+import keras
+from keras.metrics import BinaryAccuracy, Precision, Recall, F1Score
+from math import ceil
+
```
Comment (Uplink036) [inline defence/classify_detect.py]: Could do without the white space here.

File: defence/defence.py
Comment (Uplink036) [inline defence/defence.py:1]: Why is this entire file gone?

File: environment/audio_env.py
Comment (Uplink036) [inline environment/audio_env.py:1]: Interesting refactoring, not sure how i feel about it. That being said, code wise is pretty fine.

File: defence/defence.py
Comment (AdamMutzell) [reply defence/defence.py:1]: Because it is separated into classify_detect and compare_detect

File: environment/dol_env.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        reward, terminated, truncated, info = self.teach(obfuscated_audio, sr, self.reward)
+
+        next_state = preprocess_input(obfuscated_audio)
+        return next_state, reward, terminated, truncated, info
+    
+    def reward(self, transcription_similarity, audio_distance):
+        """
+        Calculate the reward based on the amount of change to the audio
+        """
+        reward = -transcription_similarity
```
Comment (Uplink036) [inline environment/dol_env.py:29]: Very bare reward function, how come?

Comment (AdamMutzell) [reply environment/dol_env.py:29]: Simply because it does not need more. Dolphin is inaudible (mostly, there is a little caveat), so no need to factor in the magnitude of the noise.

File: environment/mel_env.py
Comment (Uplink036) [inline environment/mel_env.py:1]: This i what i like with the teach thing. Less code does mean less code to maintain and go wrong.

File: models/ddpg.py [hunk 3]
```diff
@@ -172,7 +172,22 @@ def get_critic(input_size, output_size):
     model = keras.Model([state_input, action_input], outputs)
     return model
 
+def update_target(target, original, tau):
```
Comment (Uplink036) [inline models/ddpg.py:175]: Should be able to make this a no parameter function, no? Don't really see a reason why we should have inputs, when it's all from the class and things we don't change. That being said, not a must by any means.

File: environment/mel_env.py
Comment (AdamMutzell) [reply environment/mel_env.py:1]: Yep, exactly my thinking. It is more difficult to change stuff for the attacks in isolation, but hopefully these parts are almost always the same.

File: trainDOL.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+            action = agent.policy(tf_prev_state)
+           
+            state, reward, done, truncated, _ = env.step(action)
+
+            agent.buffer.record((prev_state, action, reward, state))
+            episodic_reward += reward
+            agent.learn()
+
+            update_target(agent.t_actor, agent.actor, agent.tau)
```
Comment (Uplink036) [inline trainDOL.py:60]: Does this update target truly work? Feels like it's missing an import at the very least.

File: models/ddpg.py (hunk 3 above)
Comment (AdamMutzell) [reply models/ddpg.py:175]: I have not looked into this function really. It is made by you

File: trainDOL.py (hunk 4 above)
Comment (AdamMutzell) [reply trainDOL.py:60]: Great question, I have no clue

File: defence/defence.py
Comment (Uplink036) [reply defence/defence.py:1]: Fair enough. You okay with just deleting it then?

File: trainDOL.py (hunk 4 above)
Comment (AdamMutzell) [reply trainDOL.py:60]: It is exactly the same in the other training files. I assumed it worked. Again, did not write it and nor do I understand it completely.

File: environment/dol_env.py (hunk 2 above)
Comment (Uplink036) [reply environment/dol_env.py:29]: I feel like our reward function should be able to handle / see that. I understand it's a caveat, but I feel like it _could_ make it very easy to spot as a tampered file.

File: defence/defence.py
Comment (AdamMutzell) [reply defence/defence.py:1]: There is one function that both use (get_audio_data), so keep that. It is different from other get_audio_data, so don't try and replace it. But yea, remove the commented part

File: models/ddpg.py (hunk 3 above)
Comment (Uplink036) [reply models/ddpg.py:175]: Thought it was implemented in to the class, I have been thinking about but never acted on it. Oh well, it lives in a better place now.

File: trainDOL.py (hunk 4 above)
Comment (Uplink036) [reply trainDOL.py:60]: Just to clarify, don't mean does the function itself work, but does it in the file. I feel like it's not imported, even though it should be.

File: environment/dol_env.py (hunk 2 above)
Comment (AdamMutzell) [reply environment/dol_env.py:29]: Not by ear no? It only adds frequencies which we cannot hear. So it does not matter how large the added distortion is, just that it can reduce the transcription correctness. 

The caveat is that for shorter files, these distortions actually do become audible. I think I know why and I will mention it in the report if it is necessary. I do not see this as something to be concerned about however, since the idea itself is still valid.

File: trainDOL.py (hunk 4 above)
Comment (AdamMutzell) [reply trainDOL.py:60]: It is, look at line 16

File: environment/dol_env.py (hunk 2 above)
Comment (Uplink036) [reply environment/dol_env.py:29]: No, and I suppose that is what we are trying to tackle. I was thinking more that it should work with the previous reward function, even if it mostly gives inaudible changes. As a bonus, it should limit the changes it does make and that can be useful for evading our defence, maybe.

File: trainDOL.py (hunk 4 above)
Comment (Uplink036) [reply trainDOL.py:60]: Oops. I expected it on line 11 and I guess that was just to far down.

File: environment/dol_env.py (hunk 2 above)
Comment (AdamMutzell) [reply environment/dol_env.py:29]: Thinking about the defense during attack development is expanding the scope of the project, which I don't like when it is already very large. Keeping the change down in other attacks is to make it as hard to hear as possible, not to detect with a detector no?

**Comment by Uplink036** [comment]
All of the conversations are resolved. I removed some white space and change some imports orders, but otherwise there was no problem. Merging.
